src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+               as variables in any of the {!field:rhs} which appeared in
+               the matrix that gave birth to this one. *)}
+
+  (** [pp o m] prints matrix [m] to out channel [o]. *)
+  let pp : t pp = fun oc { clauses ; _ } ->
+    let module F = Format in
+    let pp_line oc l =
+      F.fprintf oc "@[<h>" ;
+      F.pp_print_list ~pp_sep:(fun _ () -> F.fprintf oc ";@ ")
+        pp_component oc (Array.to_list l) ;
+      F.fprintf oc "@]" in
+    F.fprintf oc "{@[<v>@," ;
+    F.pp_print_list ~pp_sep:F.pp_print_cut pp_line oc
+      (List.map (fun { lhs ; _ } -> lhs) clauses) ;
+    F.fprintf oc "@.}@,"
+
+  (** [flushout_vars l] returns a mapping from position of variables into [l]
+      to the slot assigned to each variable in a {!type:term_env}. *)
+  let flushout_vars : term list -> action -> int SubMap.t = fun lhs rhs ->
+    let nvars = Bindlib.mbinder_arity rhs in
+    let rec loop found st po =
+      if found = nvars then SubMap.empty else
+      match st with
+      | [] -> SubMap.empty
+      | x :: xs ->
+         begin
+           match x with
+           | Patt(None, _, _)
+           | Vari(_)
+           | Symb(_, _)          -> loop found xs (Sub.succ po)
+           | Patt(Some(i), _, _) -> SubMap.add po i
+              (loop (succ found) xs (Sub.succ po))
+           | Appl(_, _)          -> let _, args = Basics.get_args x in
+                                   deepen found args xs po
+           | Abst(_, b)          -> let _, body = Bindlib.unbind b in
+                                   deepen found [body] xs po
+           | _                   -> assert false
+         end
+    and deepen found args remain po =
+      let argpos = loop found args (Sub.sub po) in
+      SubMap.union (fun _ _ _ -> assert false) argpos
+        (loop found remain (Sub.succ po))
+    in
+    (* The terms of the list are the subterms of the head symbol. *)
+    loop 0 lhs (Sub.succ Sub.init)
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    let r2r : Terms.rule -> rule = fun r ->
+      let variables = flushout_vars r.Terms.lhs r.Terms.rhs in
+      let term_pos = Sub.tag (Array.of_list r.Terms.lhs) in
+      { lhs = term_pos ; rhs = r.Terms.rhs
+      ; variables = variables} in
+    { clauses = List.map r2r rs ; var_catalogue = [] }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> m.clauses = []
+
+  (** [get_col n m] retrieves column [n] of matrix [m]. *)
+  let get_col : int -> t -> component list = fun ind { clauses ; _ } ->
+    List.map (fun { lhs ; _ } -> lhs.(ind)) clauses
+
+  (** [is_cons t] returns whether a term [t] is considered as a constructor. *)
+  let rec is_cons : term -> bool = function
+    | Appl(_, _) as a -> is_cons (fst (Basics.get_args a))
+    | Meta(_, _)
+    | Abst(_, _)
+    | Patt(_, _, _)   -> false
+    | Vari(_)
```

Comment:
Why Vari is a "constructor". What "constructor" means here?

Reply:
a `treecons`